src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningKeyMapper.java

```diff
@@ -0,0 +1,307 @@
+        }
+
+        static AttributeValue[] fromBytes(ScalarAttributeType hkType, ScalarAttributeType rkType, ByteBuffer buf) {
+            short hkValueLength = buf.getShort();
+            AttributeValue hkValue = fromBytes(hkType, buf, hkValueLength);
+            AttributeValue rkValue = fromBytes(rkType, buf, buf.remaining());
+            return new AttributeValue[] { hkValue, rkValue };
+        }
+
+        static AttributeValue fromBytes(ScalarAttributeType type, ByteBuffer buf) {
+            return fromBytes(type, buf, buf.limit());
+        }
+
+        private static AttributeValue fromBytes(ScalarAttributeType type, ByteBuffer buf, int size) {
+            byte[] bytes = new byte[size];
+            buf.get(bytes);
+            switch (type) {
+                case S:
+                    return new AttributeValue(new String(bytes, StandardCharsets.UTF_8));
+                case N:
+                    return new AttributeValue().withN(BigDecimalSortedBytesConverter.decode(bytes).toPlainString());
+                case B:
+                    return new AttributeValue().withB(ByteBuffer.wrap(bytes));
+                default:
+                    throw new UnsupportedOperationException("Unsupported field type: " + type);
+            }
+        }
+    }
+
+    /**
+     * Encodes BigDecimals into byte arrays such that when sorted as <b>unsigned</b> bytes, the numerical ordering is
+     * preserved.
+     * <p/>
+     * In Dynamo, a number can have up to 38 digits of precision, and can be positive, negative, or zero.
+     * <p><ul>
+     * <li>Positive range: 1E-130 to 9.9999999999999999999999999999999999999E+125
+     * <li>Negative range: -9.9999999999999999999999999999999999999E+125 to -1E-130
+     * </ul></p>
+     * (see doc: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-data-types)
+     * <p/>
+     * The format we use is SIGNUM + EXPONENT + D_1 + ... + D_n, where
+     * <p><ul>
+     * <li>SIGNUM: 0, 1, or 2, corresponding to negative, zero, or positive (1 byte)
+     * <li>EXPONENT: Represents exponent when number is in normalized notation, where the range is [0, 255], mapped to
+     * the Dynamo range [-130, 125] (1 byte)
+     * <li>D_1 to D_n: The n (at most 38) significand digits, from most significant to least (each digit takes 1 byte)
+     * <li>When the sign is negative, the exponent and significand digits' bits are flipped
+     * </ul></p>
+     */
+    @VisibleForTesting
+    static class BigDecimalSortedBytesConverter {
+
+        private static final int MIN_EXPONENT = -130;
+        private static final int MAX_EXPONENT = 125;
+
+        static byte[] encode(BigDecimal bigDecimal) {
+            bigDecimal = bigDecimal.stripTrailingZeros();
+            int precision = bigDecimal.precision();
+            byte[] byteArray = new byte[2 + precision];
+
+            // first byte represents the signum
+            byteArray[0] = toSignumByte(bigDecimal.signum());
+
```

Comment:
Moved from HashPartitioningItemMapper, with an important fix to treat bytes as unsigned, since that's how Dynamo stores them.